File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/trace/ModelTraceApplicator.java
package org.kevoree.modeling.api.trace;

import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.KActionType;
import org.kevoree.modeling.api.KObject;
import org.kevoree.modeling.api.abs.AbstractKView;
import org.kevoree.modeling.api.meta.MetaAttribute;
import org.kevoree.modeling.api.meta.MetaClass;
import org.kevoree.modeling.api.meta.MetaReference;
import org.kevoree.modeling.api.meta.MetaType;
import org.kevoree.modeling.api.rbtree.LongRBTree;
import org.kevoree.modeling.api.util.ArrayUtils;

import java.util.HashMap;
import java.util.HashSet;

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 06/08/13
 * Time: 08:54
 */
public class ModelTraceApplicator {

    private KObject _targetModel;

    public ModelTraceApplicator(KObject p_targetModel) {
        this._targetModel = p_targetModel;
    }

    public void applyTraceSequence(final TraceSequence traceSeq, final Callback<Throwable> callback) {
        try {
            final ModelTrace[] traces = traceSeq.traces();
            HashSet<Long> dependencies = new HashSet<Long>();
            for (int i = 0; i < traces.length; i++) {
                if (traces[i] instanceof ModelAddTrace) {
                    dependencies.add(((ModelAddTrace) traces[i]).paramUUID());
                    dependencies.add(traces[i].sourceUUID());
                }
                if (traces[i] instanceof ModelRemoveTrace) {
                    dependencies.add(((ModelRemoveTrace) traces[i]).paramUUID());
                    dependencies.add(traces[i].sourceUUID());
                }
                if (traces[i] instanceof ModelSetTrace) {
                    if (traces[i].meta() != null && traces[i].meta().metaType().equals(MetaType.ATTRIBUTE)) {
                        dependencies.add(traces[i].sourceUUID());
                    } else {
                        try {
                            Long paramUUID = Long.parseLong(((ModelSetTrace) traces[i]).content().toString());
                            dependencies.add(paramUUID);
                            dependencies.add(traces[i].sourceUUID());
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
            _targetModel.view().lookupAll(ArrayUtils.flatSet(dependencies)).then(new Callback<KObject[]>() {
                @Override
                public void on(KObject[] kObjects) {
                    HashMap<Long, KObject> cached = new HashMap<Long, KObject>();
                    for (int i = 0; i < traces.length; i++) {
                        try {
                            ModelTrace trace = traces[i];
                            KObject sourceObject = cached.get(trace.sourceUUID());
                            if (sourceObject != null) {
                                if (trace instanceof ModelRemoveTrace) {
                                    ModelRemoveTrace removeTrace = (ModelRemoveTrace) trace;
                                    KObject param = cached.get(removeTrace.paramUUID());
                                    if (param != null) {
                                        sourceObject.mutate(KActionType.REMOVE, (MetaReference) removeTrace.meta(), param);
                                    }
                                } else if (trace instanceof ModelAddTrace) {
                                    ModelAddTrace addTrace = (ModelAddTrace) trace;
                                    KObject param = cached.get(addTrace.paramUUID());
                                    if (param != null) {
                                        sourceObject.mutate(KActionType.ADD, (MetaReference) addTrace.meta(), param);
                                    }
                                } else if (trace instanceof ModelSetTrace) {
                                    ModelSetTrace setTrace = (ModelSetTrace) trace;
                                    if (trace.meta() != null && trace.meta().metaType().equals(MetaType.ATTRIBUTE)) {
                                        sourceObject.set((MetaAttribute) trace.meta(), setTrace.content());
                                    } else {
                                        try {
                                            Long paramUUID = Long.parseLong(((ModelSetTrace) traces[i]).content().toString());
                                            KObject param = cached.get(paramUUID);
                                            if (param != null) {
                                                sourceObject.mutate(KActionType.SET, (MetaReference) trace.meta(), param);
                                            }
                                        } catch (Exception e) {
                                            e.printStackTrace();
                                        }
                                    }
                                } else if (trace instanceof ModelNewTrace) {
                                    LongRBTree universeTree = new LongRBTree();
                                    universeTree.insert(_targetModel.universe().key(), _targetModel.now());
                                    KObject newCreated = ((AbstractKView) _targetModel.view()).createProxy((MetaClass) trace.meta(), universeTree, trace.sourceUUID());
                                    _targetModel.universe().model().manager().initKObject(newCreated, _targetModel.view());
                                    cached.put(newCreated.uuid(), newCreated);
                                } else {
                                    System.err.println("Unknow traceType: " + trace);
                                }
                            } else {
                                System.err.println("Unknow object: " + trace);
                            }
                        } catch (Exception e) {
                            e.printStackTrace();
                            System.err.println("Error " + e);
                        }
                    }
                    callback.on(null);
                }
            });
        } catch (Exception e) {
            callback.on(e);
        }
    }

}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/trace/ModelTraceApplicator.kt
package org.kevoree.modeling.api.trace

import org.kevoree.modeling.api.util.ActionType
import org.kevoree.modeling.api.KMFContainer
import org.kevoree.modeling.api.KMFFactory

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 06/08/13
 * Time: 08:54
 */
class ModelTraceApplicator(val targetModel: KMFContainer, val factory: KMFFactory) {

    var pendingObj: KMFContainer? = null
    var pendingParent: KMFContainer? = null
    var pendingParentRefName: String? = null
    var pendingObjPath: String? = null
    var fireEvents: Boolean = true

    private fun tryClosePending(srcPath: String?) {
        if(pendingObj != null && pendingObjPath != srcPath){
            pendingParent!!.reflexiveMutator(ActionType.ADD, pendingParentRefName!!, pendingObj, true, fireEvents)
            pendingObj = null
            pendingObjPath = null
            pendingParentRefName = null
            pendingParent = null
        }
    }

    public fun createOrAdd(previousPath: String?, target: KMFContainer, refName: String, potentialTypeName: String?) {
        var targetElem: Any? = null
        if(previousPath!=null){
            targetElem = targetModel.findByPath(previousPath)
        }
        if(targetElem != null){
            target.reflexiveMutator(ActionType.ADD, refName, targetElem, true, fireEvents)
        } else {
            //add to pending
            if(potentialTypeName == null){
                throw Exception("Unknow typeName for potential path $previousPath, to store in $refName, unconsistency error")
            }
            pendingObj = factory.create(potentialTypeName)
            pendingObjPath = previousPath;
            pendingParentRefName = refName;
            pendingParent = target;
        }
    }

    public fun applyTraceOnModel(traceSeq: TraceSequence) {
        for(trace in traceSeq.traces){
            var target: KMFContainer = targetModel;
            if(trace is ModelAddTrace){
                tryClosePending(null);
                if(trace.srcPath != ""){
                    var resolvedTarget = targetModel.findByPath(trace.srcPath)
                    if(resolvedTarget==null){
                        throw Exception("Add Trace source not found for path : " + trace.srcPath + " pending " + pendingObjPath + "\n" + trace.toString())
                    }
                    target = resolvedTarget!!
                }
                createOrAdd(trace.previousPath, target, trace.refName, trace.typeName)
            }
            if(trace is ModelAddAllTrace){
                tryClosePending(null);
                var i = 0
                for(path in trace.previousPath!!){
                    createOrAdd(path, target, trace.refName, trace.typeName!!.get(i))
                    i++
                }
            }
            if(trace is ModelRemoveTrace){
                tryClosePending(trace.srcPath);
                var tempTarget: KMFContainer? = targetModel;
                if(trace.srcPath != ""){
                    tempTarget = targetModel.findByPath(trace.srcPath) as? KMFContainer;
                }
                if(tempTarget != null){
                    //Potentially null if top tree already dropped
                    tempTarget!!.reflexiveMutator(ActionType.REMOVE, trace.refName, targetModel.findByPath(trace.objPath), true, fireEvents)
                }
            }
            if(trace is ModelRemoveAllTrace){
                tryClosePending(trace.srcPath);
                var tempTarget: KMFContainer? = targetModel;
                if(trace.srcPath != ""){
                    tempTarget = targetModel.findByPath(trace.srcPath) as? KMFContainer;
                }
                if(tempTarget != null){
                    tempTarget!!.reflexiveMutator(ActionType.REMOVE_ALL, trace.refName, null, true, fireEvents)
                }
            }
            if(trace is ModelSetTrace){
                tryClosePending(trace.srcPath);
                if(trace.srcPath != "" && trace.srcPath != pendingObjPath){
                    var tempObject = targetModel.findByPath(trace.srcPath)
                    if(tempObject == null){
                        throw Exception("Set Trace source not found for path : " + trace.srcPath + " pending " + pendingObjPath + "\n" + trace.toString())
                    }
                    target = tempObject as KMFContainer;
                } else {
                    if(trace.srcPath == pendingObjPath && pendingObj != null){
                        target = pendingObj!!;
                    }
                }
                if(trace.content != null){
                    target.reflexiveMutator(ActionType.SET, trace.refName, trace.content, true, fireEvents)
                } else {
                    var targetContentPath: Any? = if(trace.objPath != null){
                        targetModel.findByPath(trace.objPath)
                    } else {
                        null
                    };
                    if(targetContentPath != null){
                        target.reflexiveMutator(ActionType.SET, trace.refName, targetContentPath, true, fireEvents)
                    } else {
                        if(trace.typeName != null && trace.typeName != ""){
                            createOrAdd(trace.objPath, target, trace.refName, trace.typeName) //must create the pending element
                        } else {
                            target.reflexiveMutator(ActionType.SET, trace.refName, targetContentPath, true, fireEvents) //case real null content
                        }
                    }
                }
            }
        }
        tryClosePending(null);
    }

}
